### ISSUE TITLE
Crash in tranform_keys Presto lambda function

### ISSUE THREAD

**Issue by mbasmanova** [facebookincubator/velox#9288]
### Bug description

A simple repro:

```
TEST_F(TransformKeysTest, evaluateSubsetOfRows) {
  // Test to verify that output complex vector of valid internal state is
  // generated when only a subset of the rows are evaluated. To simulate this,
  // the trailing rows are unselected to generate a keys vector of smaller size
  // so that indices for those rows would point to out of bounds location in
  // keys.
  vector_size_t size = 100;
  auto input = makeRowVector({
      makeMapVector<int64_t, int32_t>(
          size,
          [](auto row) { return row % 5; },
          [](auto row) { return row % 7; },
          [](auto row) { return row % 11; },
          nullEvery(13)),
  });

  // 33 (size / 3) bits are used, but a total of 64 bits are selected. There is
  // a tail of 31 selected bits.
  SelectivityVector inputRows(size / 3);

  auto result = evaluate<MapVector>(
      "transform_keys(c0, (k, v) -> k + 5)", input, inputRows);
  auto expectedResult = makeMapVector<int64_t, int32_t>(
      inputRows.size(),
      [](auto row) { return row % 5; },
      [](auto row) { return row % 7 + 5; },
      [](auto row) { return row % 11; },
      nullEvery(13));
  assertEqualVectors(expectedResult, result, inputRows);
}
```

TransformKeys::apply receives a map vector of size 100 with a SelectivityVector of size 33 and all rows [0, 32] selected. It evaluates lambda on 33 rows and produces a new keys vector of size 33. It then creates a new map vector of size 100 with keys vector of size 33. It uses addNullsForUnselectedRows to mark all rows outside of [0, 32] as null. 

This should work, but addNullsForUnselectedRows fails to properly set rows outside of [0, 32] to null. It uses memcpy to copy SelectivityVectors bits into nulls buffer. It copies a few extra rows at the tail because 33 is not a multiple of 8. If these bits happen to be 1 (selected), the null buffer gets a few rows past row 32 marked as not-null. This results in an invalid MapVector and causes a crash in MapVector::canonicalize. 

```
  memcpy(
      nulls->asMutable<uint64_t>(),
      rows.asRange().bits(),
      bits::nbytes(rows.end()));
```

A fix is to replace memcpy with bits::orBits: #9279

Also, there is no need to create MapVector of the original size. It is sufficient to create a vector of rows.end() size. Similarly, no need to allocate nulls buffer to fit original size bits. This optimization is in #9280 

There are a few more places that use memcpy incorrectly. Need to fix these as well.

- ZipWith.cpp
- ArrayAggAggregate.cpp
- MapAggAggregate.cpp
- MapUnionAggregate.cpp
- SetAggregates.cpp

CC: @bikramSingh91 

### System information

n/a

**Comment by mbasmanova** [comment]
Usage of memcpy in the following places is ok. These places create nulls buffers with N entries and copy bits::nbytes(N) bytes. The extra tail bits are not valid in the nulls buffer and therefore their values do not matter. This is different from transformXxx functions where nulls buffers have N valid entries, but we copy bits::nbytes(M) bytes into it where M could be less than N in case rows after M have an error. In this case tail bits may fall within the valid range of the nulls buffer.

- ZipWith.cpp
- ArrayAggAggregate.cpp
- MapAggAggregate.cpp
- MapUnionAggregate.cpp
- SetAggregates.cpp

That said, it would be nice to refactor this code to move the logic of creating nulls buffer with nulls for unselected rows into a shared place (a method on SelectivityVector?).

**Comment by mbasmanova** [comment]
Fixed in #9279
